domain/model/state/state.go

```diff
@@ -566,17 +619,79 @@ AND owner_name = ?
 		)
 	}
 
-	if !credentialName.Valid {
-		return cloudName, credential.Key{}, nil
+	return cloudName, credentialKey, nil
+}
+
+// NamespaceForModel returns the database namespace that is provisioned for a
+// model id. If no model is found for the given id then a [modelerrors.NotFound]
+// error is returned. If no namespace has been provisioned for the model then a
+// [modelerrors.ModelNamespaceNotFound] error is returned.
+func (st *State) NamespaceForModel(ctx context.Context, id coremodel.UUID) (string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	q := `
+    SELECT m.uuid, mn.namespace
+    FROM model m
+	LEFT JOIN model_namespace mn ON m.uuid = mn.model_uuid
+    WHERE m.uuid = ?
+`
+	var namespace sql.NullString
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		return tx.QueryRowContext(ctx, q, id).Scan(&id, &namespace)
+	})
+
+	if errors.Is(err, sql.ErrNoRows) {
+		return "", fmt.Errorf("%w for id %q", modelerrors.NotFound, id)
+	} else if err != nil {
+		return "", fmt.Errorf(
+			"getting database namespace for model %q: %w",
+			id,
+			domain.CoerceError(err),
+		)
+	}
+
+	if !namespace.Valid {
+		return "", fmt.Errorf(
+			"%w for id %q",
+			modelerrors.ModelNamespaceNotFound,
+			id,
+		)
+	}
+
+	return namespace.String, nil
+}
+
+// registerModelNamespace is responsible for taking a constructed model and
+// registering a new DQlite namespace for the model. If no model is found the
+// provided uuid an error satisfying [modelerrors.NotFound] is returned.
+func registerModelNamespace(
+	ctx context.Context,
+	tx *sql.Tx,
+	uuid coremodel.UUID,
+) (string, error) {
+	q := "INSERT INTO namespace_list (namespace) VALUES (?)"
+
+	_, err := tx.ExecContext(ctx, q, uuid.String())
+	if jujudb.IsErrConstraintPrimaryKey(err) {
```

Comment:
So we need to avoid prematurely upgrading the txn to write by first selecting for existing row and returning an error if necessary, rather than attempting to write and handling the primary key error. This is for performance reasons with dqlite, where it is very expensive to create write txns.

Reply:
As discussed this is fine. Chances of this happening are nil and this is defensive programming.